File: src/mandelbrot.c
#include <stdio.h>
#include <stdlib.h>

unsigned int iterate(unsigned int max_modulus, unsigned int max_iterations, double cr, double ci) {
    double zr, zi, rr, ri, rc;
    unsigned int i;

    zr = zi = 0.0;

    i = 0;

    do {
        rr = zr * zr;
        ri = zi * zi;
        rc = zr * zi;

        zr = rr - ri + cr;
        zi = 2 * rc + ci;

        i++;

        if (rr + ri > max_modulus) return i;
        if (i > max_iterations) return 0;
    } while (1);
}

void mandelbrot_draw(unsigned int max_modulus, unsigned int max_iterations, unsigned int resolution) {
    double cr, ci;
    unsigned int unit, result, width, height;
    unsigned int i, j;

    unit = resolution / 6;
    width = unit * 3 * 2;
    height = unit * 2;

    for (j = 0; j <= height; j++) {
        for (i = 0; i <= width; i++) {
            cr = ((double) i / (double) unit) / 2.0 - 2.0;
            ci = 1.0 - (double) j / (double) unit;

            result = iterate(max_modulus, max_iterations, cr, ci) % 26;

            putchar(result == 0 ? ' ' : 64 + result);
        }

        putchar('\n');
    }
}

int main() {
    unsigned int max_modulus;
    unsigned int max_iterations;
    unsigned int resolution;

    max_modulus = atoi(getenv("MANDELBROT_MAX_MODULUS"));
    max_iterations = atoi(getenv("MANDELBROT_MAX_ITERATIONS"));
    resolution = atoi(getenv("MANDELBROT_RESOLUTION"));

    mandelbrot_draw(max_modulus, max_iterations, resolution);

    printf("Max modulus: %d\n", max_modulus);
    printf("Max iterations: %d\n", max_iterations);
    printf("Resolution: %d\n", resolution);

    return 0;
}

File: src/mandelbrot.go
package main

import (
    "fmt"
    "os"
    "strconv"
)

func iterate(max_modulus int, max_iterations int, ci float64, cr float64) int {
    var zr, zi = 0.0, 0.0
    var i = 0
    var result int

    for true {
        rr := zr * zr
        ri := zi * zi
        rc := zr * zi

        zr = rr - ri + cr
        zi = 2*rc + ci

        i++

        if int(rr+ri) > max_modulus {
            result = i
            break
        }

        if i > max_iterations {
            result = 0
            break
        }
    }

    return result
}

func mandelbrot_draw(max_modulus int, max_iterations int, resolution int) {
    var cr, ci float64
    var unit = resolution / 6
    var width = unit * 3 * 2
    var height = unit * 2
    var result int
    var i, j int

    for j = 0; j <= height; j++ {
        for i = 0; i <= width; i++ {
            ci = ((float64(i) / float64(unit)) / 2.0) - 2.0
            cr = 1.0 - (float64(j) / float64(unit))

            result = iterate(max_modulus, max_iterations, cr, ci) % 26

            if result == 0 {
                fmt.Print(" ")
            } else {
                fmt.Print(string(64 + result))
            }
        }

        fmt.Print("\n")
    }
}

func main() {
    var max_modulus, _ = strconv.Atoi(os.Getenv("MANDELBROT_MAX_MODULUS"))
    var max_iterations, _ = strconv.Atoi(os.Getenv("MANDELBROT_MAX_ITERATIONS"))
    var resolution, _ = strconv.Atoi(os.Getenv("MANDELBROT_RESOLUTION"))

    mandelbrot_draw(max_modulus, max_iterations, resolution)

    fmt.Println("Max modulus: ", max_modulus)
    fmt.Println("Max iterations: ", max_iterations)
    fmt.Println("Resolution: ", resolution)
}
